### PR TITLE
Westernization & literacy

### PR DESCRIPTION
**Pull request by Zemurin**
- Westernization mapper for eurocentrism
- Determining if a nation is westernized or not (vanilla + eurocentrism)
- Literacy calc + adjustments (literacy is an INTERNAL score that's yet to be mapped to Vic3 Literacy Effects!)

### CODE REVIEW

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 1]
```diff
@@ -351,3 +352,132 @@ std::string V3::Country::getAdjective(const std::string& language) const
...
+
+	const auto currentYear = std::fmax(lastDate.diffInYears(date("0.1.1")), 0);
+	const auto yearFactor = (0.1 + 4'614'700 * currentYear) / (1 + static_cast<double>(103'810'000.0f) * currentYear - 54'029 * pow(currentYear, 2));
+
+	processedData.literacy *= yearFactor;
+
+	// Apply cultural mod.
+	processedData.literacy *= 1 + (static_cast<double>(cultureMapper.getLiteracyScoreForCulture(*processedData.cultures.begin())) - 5.0) * 10.0 / 100.0;
+
+	// Reduce the literacy for non-resternized nations according to their civLevel score.
```
Comment (gawquon) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:396]: typo
```suggestion
	// Reduce the literacy for non-westernized nations according to their civLevel score.
```

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 2]
```diff
@@ -351,3 +352,132 @@ std::string V3::Country::getAdjective(const std::string& language) const
...
+	 double topInstitutions,
+	 const mappers::CultureMapper& cultureMapper,
+	 Configuration::EUROCENTRISM eurocentrism)
+{
+	// This is base calc, from EU4. Even western countries in severe tech deficit will have a lower civLevel score.
+	const auto totalTechs = sourceCountry->getMilTech() + sourceCountry->getAdmTech() + sourceCountry->getDipTech();
+	processedData.civLevel = (totalTechs + 31.0 - topTech) * 4;
+	processedData.civLevel += (static_cast<double>(sourceCountry->getNumEmbracedInstitutions()) - topInstitutions) * 8;
+
+	// If we're eurocentric, we're *ignoring* thech deficits and artificially deflating whatever score was acheived earlier for non-western countries.
```
Comment (gawquon) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:412]: typo
```suggestion
	// If we're eurocentric, we're *ignoring* tech deficits and artificially deflating whatever score was achieved earlier for non-western countries.
```

File: EU4toV3Tests/MapperTests/CultureMapperTests/WesternizationMapperTests/WesternizationMappingTests.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	std::stringstream input;
+	const mappers::WesternizationMapping mapping(input);
+
+	EXPECT_EQ(5, mapping.getWesternization());
+	EXPECT_EQ(5, mapping.getLiteracy());
+	EXPECT_EQ(5, mapping.getIndustry());
+	EXPECT_TRUE(mapping.getTraits().empty());
+}
+
+TEST(Mappers_WesternizationMappingTests, mappjngCanBeLoaded)
```
Comment (Idhrendur) [inline EU4toV3Tests/MapperTests/CultureMapperTests/WesternizationMapperTests/WesternizationMappingTests.cpp:16]:
```suggestion
TEST(Mappers_WesternizationMappingTests, mappingCanBeLoaded)
```

**Review by Idhrendur — APPROVED** [review]

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 4]
```diff
@@ -351,3 +352,132 @@ std::string V3::Country::getAdjective(const std::string& language) const
...
+		1490: 0.2
+		1350: 0.15
+		It will fail to hit those points exactly but won't err by much.
+	*/
+
+	auto lastDate = datingData.lastEU4Date;
+	if (lastDate > datingData.hardEndingDate)
+		lastDate = datingData.hardEndingDate;
+
+	const auto currentYear = std::fmax(lastDate.diffInYears(date("0.1.1")), 0);
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:388]: Sure, moved to a function.

**Pull request merged**